text/0113-virtual-script-container.md

```diff
@@ -0,0 +1,213 @@
+champion: Caridy Patiño, Ted Conn
+---
+
+# Virtual Script Container for LWC Apps
+
+## Summary
+
+We need a way for application developers to integrate common third-party libraries into their Web Component based apps (specifically, LWC applications). These libraries, such as Google Analytics, are severely hindered by the Shadow DOM which prevents global interaction with any element on the page.
+
+By providing a virtual container for scripts to run, see, and interact with all the Shadow DOM trees as if it were the light DOM, application developers can continue to run their existing integrations until these libraries support Shadow DOM semantics natively.
+
+## Basic example
+
+```html
+<virtual-script src="//cdn.optimizely.com/js/12345678.js"></virtual-script>
+```
+
+`virtual-script` is a custom element which acts the same as the `script` tag, but any code which is evaluated inside this container is able to traverse the entire Shadow  DOM tree using light DOM semantics. So if the above script runs `document.querySelectorAll('button')`, it will return all `<button>` elements regardless if they are encapsulated inside a `ShadowRoot` or not.
+
+```html
+<virtual-script>
+    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
+    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
+    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
+
+    ga('create', 'UA-XXXXX-Y', 'auto');
+    ga('send', 'pageview');
+</virtual-script>
+```
+
+Inline scripting would also be supported, such as the above example which uses script injection to fetch its required resources.
+
+## Motivation
+
+Web component based applications need the ability to integrate with third party libraries. Because
+web components rely on the Shadow DOM for encapsulation, this development paradigm does not work
+with libraries who are expecting to interact with the application globally. Before there was
+component encapsulation via Shadow DOM it was possible for library authors to:
+
+- Listen for events and know who fired them
+- Define global styling rules that would get applied on every element
+- Query the entire document for any element on the page
+
+These libraries are typically added as global scripts to the document root rather than deeper 
+integrations into the components themselves. When every component in the application has a ShadowRoot attached to them, these libraries will not work as expected.
+
+Some examples of these libraries are:
+
+- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
+- **Personalization platforms**: Optimizely, Evergage, Google Optimize
+- **Instrumentation tools**: New Relic, Sentry.io
+
+The goal of this proposal is to provide a way for application developers to continue to use their existing integrations until library authors support Shadow DOM integration.
```

Comment:
This makes sense to enable adoption. It also raises some significant concerns. Once this is released, developers can write any code they want and load that code using a `<virtual-script>` element to bypass the encapsulation of each web component. This a free escape hatch which breaks the principle of encapsulation by its very design. That opens the door to developers using this intentionally to write code which relies on implementation details inside a web component.

1) What is the planned support statement for using `<virtual-script>` when code it executes "breaks" when changes occur to the application around it?
2) Are there plans to support `<virtual-script>` on Salesforce platform?

Reply:
> That opens the door to developers using this intentionally to write code which relies on implementation details inside a web component.

Developers can already bypass that by traversing the shadow roots, right?

> What is the planned support statement for using <virtual-script> when code it executes "breaks" when changes occur to the application around it?

That is a great question and something @gregwhitworth was alluding to before, but we decided that discussion was outside of the scope of this RFC, and better suited for a security assessment when the usage of `virtual-script` is allowed in a Salesforce product.

> Are there plans to support <virtual-script> on Salesforce platform?

Currently only B2C experiences will support this element, and we will have a corresponding threat analysis. That being said, [we already allow Locker Disablement](https://developer.salesforce.com/docs/atlas.en-us.communities_dev.meta/communities_dev/communities_dev_security_locker.htm) and B2C experiences don't share the same security concerns as the rest of the platform.